packages/data/src/user-preferences/use-user-preferences.js

```diff
@@ -91,6 +91,15 @@ export const useUserPreferences = () => {
 					};
 				}
 
+				// Optimistically propagate new woocommerce_meta to the store for instant update.
+				receiveCurrentUser( {
+					...user,
+					woocommerce_meta: {
+						...user.woocommerce_meta,
+						...metaData,
+					},
+				} );
+
 				// Use saveUser() to update WooCommerce meta values.
 				const updatedUser = await saveUser( {
```

Comment:
Would it be possible to only call `receiveCurrentUser` once and not again with `receiveCurrentUser( updatedUser )` once updated if no errors are encountered?

It might also be nice (in a follow-up possibly) to do some more error handling and create a notice if preferences aren't being saved.

Reply:
Good catch @joshuatf !

It definitely makes sense not to save twice due to the possibility of fast clicking. I removed that in https://github.com/woocommerce/woocommerce-admin/pull/4637/commits/11fb920dc4f97171f3f3c59572ee0e91a8d05600.

As it stands now, if an error occurs its returned from `updateUserPreferences` so an async function can choose to handle it however it wants.

Alternatively, we can fire off a core notice if there was an error. The UI would update, and a notice would appear once the error is returned.

Reply:
>As it stands now, if an error occurs its returned from updateUserPreferences so an async function can choose to handle it however it wants.

I do like the idea of a component choosing how it handles error messaging.

Maybe one piece of error handling that belongs in the store is reverting the options with `receiveCurrentUser()` if an error does occur.  And then continuing to pass the error message along so the component can choose how to handle the UI aspect of it.

But this is hopefully an edge case and I'm good if we handle this in a follow-up.